Replace FreeImage dependency with stb

wentasah opened the pull request:
# 🦟 Bug fix

Fixes #388

## Summary
This removes FreeImage dependency. This library is full of security vulnerabilities [1] [2] and distributions such as NixOS and Arch no longer ship it, making it difficult to build Gazebo there. It's not an option to replace FreeImage with its fork called FreeImageRe [3], which fixes some of the vulnerabilities, because many of them still seem to be present. Therefore, we replace FreeImage with the stb project [4], which seems to have better security track record [5]. We vendor the header files from the project from its commit f1c79c0 ("Merge pull request 1851 from jeffrbig2/master", 2025-10-25).

With this implementation all tests pass, but it is quite likely that some things will behave differently. Likely the support of 16-bit or HDR (32-bit float) images will behave differently. For example, PNG saving is supported only for 8-bit images.

[1]: https://security-tracker.debian.org/tracker/source-package/freeimage
[2]: https://www.cvedetails.com/vendor/15676/
[3]: https://github.com/agruzdev/FreeImageRe
[4]: https://github.com/nothings/stb
[5]: https://www.cvedetails.com/vendor/19263/Stb-Project.html

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

## Test it

All `gz-common` tests pass. I also run tests of other `gz-*` packages and didn't observe any new failure (it seems that at least some `gz-rendering` and `gz-sim` tests use functionality from `Image.cc`.

I run `gz sim` with some example worlds. Taking screenshots uses `Image.cc` to save it. `thermal_camera.sdf` also uses `Image.cc`, perhaps to show images of the camera. The images look correct.

wentasah committed:
Replace FreeImage dependency with stb

This removes FreeImage dependency. This library is full of security
vulnerabilities [1][2] and distributions such as NixOS and Arch no
longer ship it, making it difficult to build Gazebo there. It's not an
option to replace FreeImage with its fork called FreeImageRe [3],
which fixes some of the vulnerabilities, because many of them still
seem to be present. Therefore, we replace FreeImage with the stb
project [4], which seems to have better security track record [5]. We
vendor the header files from the project from its commit
f1c79c0 ("Merge pull request 1851 from jeffrbig2/master", 2025-10-25).

With this implementation all tests pass, but it is quite likely that
some things will behave differently. Likely the support of 16-bit or
HDR (32-bit float) images will behave differently. For example, PNG
saving is supported only for 8-bit images.

[1]: https://security-tracker.debian.org/tracker/source-package/freeimage
[2]: https://www.cvedetails.com/vendor/15676/
[3]: https://github.com/agruzdev/FreeImageRe
[4]: https://github.com/nothings/stb
[5]: https://www.cvedetails.com/vendor/19263/Stb-Project.html

Closes #388

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

iche033 commented:
Thanks for the contribution! There was an attempt at integrating stb in https://github.com/gazebosim/gz-common/pull/590. That PR added stb as an additional image loader while still keeping freeimage. I think that PR was closed because we just didn't have time to take it to the finish line?  cc @mjcarroll.

Note this PR removes a dependency in a stable release version which I think we typically don't do. We should retarget this to the `main` branch first and discuss whether it's feasible to backport to the release branches.

scpeters reviewed (commented):
the `libfreeimage-dev` package is also listed in [.github/ci/packages.apt](https://github.com/wentasah/gz-common/blob/replace-freeimage-gz7/.github/ci/packages.apt#L6) for use in CI, and that can be removed

I will try to find a maintainer to prioritize review of this PR

scpeters committed:
Merge branch 'main' into replace-freeimage-gz7

scpeters commented:
testing downstream packages against this branch using https://github.com/gazebo-tooling/gazebodistro/commit/092229472d95c05ba544d33b7640ad5be501e160:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ci__colcon_any-manual_ubuntu_noble_amd64&build=58)](https://build.osrfoundation.org/job/ci__colcon_any-manual_ubuntu_noble_amd64/58/) https://build.osrfoundation.org/job/ci__colcon_any-manual_ubuntu_noble_amd64/58/

Commit:
Move STB to subdirectory

This allows us to place the LICENSE file in the directory and aligns
with other vendored dependencies

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>

Commit:
Update bazel for STB dependency

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>

Commit:
Remove freeimage from packages.apt

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>

Commit:
Lint

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>

mjcarroll committed:
Ignore deprecated snprintf on Apple

Signed-off-by: Michael Carroll <carroll.michael@gmail.com>

mjcarroll committed:
Suppress changes in image_write not image_resize

Signed-off-by: Michael Carroll <carroll.michael@gmail.com>

Commit:
Missing #endif

Signed-off-by: Michael Carroll <mjcarroll@intrinsic.ai>

Commit:
Merge branch 'main' into replace-freeimage-gz7

scpeters reviewed (commented):
can you add a `graphics/src/STB/README` file with info about the specific commit that this is vendored from? It can be similar to https://github.com/gazebosim/gz-utils/blob/0958c36cdd0ac9d9fc594ef4436f323eddd593c6/cli/README.md?plain=1#L4

wentasah committed:
Add STB/README with reference to vendored commit

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

wentasah committed:
Try to suppress sprintf deprecation warning on Macs

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

wentasah committed:
Try to fix windows link error

See https://github.com/gazebosim/gz-common/pull/725#discussion_r2769447653.

Co-authored-by: Jose Luis Rivero <jrivero@osrfoundation.org>
Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

scpeters committed:
Merge branch 'main' into replace-freeimage-gz7

scpeters reviewed (approved):
I haven't reviewed all the changes in Image.cc, but the CI results look ok

Commit:
Merge branch 'main' into replace-freeimage-gz7

wentasah committed:
Merge remote-tracking branch 'origin/main' into replace-freeimage-gz7

This fixes conflict in Image::ChannelData. The resolution was taken
from my original version of this PR, which targeted gz-common7.

azeey commented on graphics/src/Image.cc:
```diff
@@ -670,71 +590,42 @@ Image::PixelFormatType Image::ConvertPixelFormat(const std::string &_format)
+    }
+    case 32: {
+      auto * pixel = reinterpret_cast<const float *>(bitmap) + i * ch;
+      data[i] = pixel[ch_i] * 255.0f;
```
This is causing a warning. Can you cast it to `unsigned char`?

wentasah committed:
Try to suppress warning on Windows

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

azeey commented on graphics/src/Image.cc:
```diff
@@ -170,313 +173,280 @@ int Image::Load(const std::string &_filename)
+    gzerr << "Image: Unsupported number of channels [" << channels << "] \n";
+  }
+}
+math::Color Image::Pixel(unsigned int _x, unsigned int _y) const {
```
Please revert all the style changes to minimize the diff size.
```suggestion
//////////////////////////////////////////////////
math::Color Image::Pixel(unsigned int _x, unsigned int _y) const
{
```

azeey reviewed (commented):
Looks great! Thanks for the awesome contribution. I've left a few comments/questions.

wentasah committed:
Follow the style guide and minimize diff

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

wentasah commented on graphics/src/Image.cc:
```diff
@@ -170,313 +173,280 @@ int Image::Load(const std::string &_filename)
+  auto buffer = reinterpret_cast<std::vector<unsigned char> *>(context);
+  auto prev_size = buffer->size();
+  buffer->resize(prev_size + size);
+  std::memcpy(buffer->data() + prev_size, data, size);
```
This code is a callback called multiple times by `stbi_write_png_to_func()` to append data to the PNG buffer. We need to keep the content added by previous calls.

azeey replied:
I see. Thanks for the explanation.

azeey reviewed (commented):
There's a linter error and one more comment. Otherwise, LGTM!

wentasah committed:
Wrap lines longer than 80 characters

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

wentasah committed:
Don't destroy original image if rescaling fails

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

wentasah committed:
Make Image::SetFromData partially handle Bayer pixel format

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

Commit:
Merge branch 'main' into replace-freeimage-gz7

wentasah committed:
Add support for BGR_INT8 format in Image::SetFromData

Signed-off-by: Michal Sojka <michal.sojka@cvut.cz>

azeey reviewed (approved):
Great work!! Just wait for green CI now

Also running CI across all Gazebo libraries with this branch:
[![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ci__colcon_gpu_any-manual_ubuntu_noble_amd64&build=19)](https://build.osrfoundation.org/view/ci/job/ci__colcon_gpu_any-manual_ubuntu_noble_amd64/19/)

azeey commented:
Looks like there are a few infra related issues with https://build.osrfoundation.org/view/ci/job/ci__colcon_gpu_any-manual_ubuntu_noble_amd64/19/. Non of the GUI tests were able to run because there was no display available. DART is also not being installed properly. CC @j-rivero 

I'll go ahead and merge this and we'll check the nightlies for any regressions.

taylorhoward92 commented:
> Thanks for the contribution! There was an attempt at integrating stb in #590. That PR added stb as an additional image loader while still keeping freeimage. I think that PR was closed because we just didn't have time to take it to the finish line? cc @mjcarroll.
> 
> Note this PR removes a dependency in a stable release version which I think we typically don't do. We should retarget this to the `main` branch first and discuss whether it's feasible to backport to the release branches.

Has there been any discussion about if this is feasible to back port to release branches? I couldn't see any related discussion in this PR.

scpeters commented:
> > Note this PR removes a dependency in a stable release version which I think we typically don't do. We should retarget this to the `main` branch first and discuss whether it's feasible to backport to the release branches.
> 
> Has there been any discussion about if this is feasible to back port to release branches? I couldn't see any related discussion in this PR.

I think migrating away from `freeimage` is needed soonest on macOS since `freeimage` has been deprecated in homebrew-core (see https://github.com/osrf/homebrew-simulation/issues/3298), and after merging this PR I realized that our ogre1.9 and ogre2.3 formulae are still using `freeimage` which led me to add a `gz-rotary-ogre2.3-vendor` formula in https://github.com/osrf/homebrew-simulation/pull/3391 and experiment with using it with gz-rendering in https://github.com/osrf/homebrew-simulation/pull/3394. Some things don't render the same (see https://github.com/osrf/homebrew-simulation/pull/3394#issuecomment-4101535199). I think we should prioritize getting a fully working combination of Gazebo / ogre without `freeimage` on `main` before backporting.

If you are able to test https://github.com/osrf/homebrew-simulation/pull/3394 and help identify the underlying issues there, please do so! That would help us move this forward

wentasah commented:
> I think we should prioritize getting a fully working combination of Gazebo / ogre without `freeimage` on `main` before backporting.

BTW, in nix-ros-overlay, we have the needed patches backported to gz-common-vendor in all supported ROS distros. If you decide to backport, feel free to ping me and I can submit the PRs. There were some conflicts to resolve so it's not necessary to resolve them again.

scpeters commented:
> > I think we should prioritize getting a fully working combination of Gazebo / ogre without `freeimage` on `main` before backporting.
> 
> BTW, in nix-ros-overlay, we have the needed patches backported to gz-common-vendor in all supported ROS distros. If you decide to backport, feel free to ping me and I can submit the PRs. There were some conflicts to resolve so it's not necessary to resolve them again.

we just finished removing `freeimage` from rotary (`main` branches) and are considering backporting to Jetty. @wentasah do you mind opening these backport PR's? Thank you!

wentasah commented:
@scpeters Done in #803. Let me know if you're also interested in PRs for older versions.
